featuremanagement/_featuremanagerbase.py

```diff
@@ -225,31 +226,74 @@ def _build_targeting_context(self, args: Tuple[Any]) -> TargetingContext:
-                    return
-                evaluation_event.reason = VariantAssignmentReason.DEFAULT_WHEN_DISABLED
-                return
-            if not evaluation_event.enabled:
-                FeatureManagerBase._check_default_disabled_variant(evaluation_event)
-                evaluation_event.variant = self._variant_name_to_variant(
-                    feature_flag, feature_flag.allocation.default_when_disabled
-                )
-                return
 
-            self._assign_variant(feature_flag, targeting_context, evaluation_event)
+        if not feature_flag.variants:
+            return
+
+        if not feature_flag.allocation:
+            evaluation_event.reason = (
+                VariantAssignmentReason.DEFAULT_WHEN_ENABLED
+                if evaluation_event.enabled
+                else VariantAssignmentReason.DEFAULT_WHEN_DISABLED
+            )
+            return
+        if not evaluation_event.enabled:
+            FeatureManagerBase._check_default_disabled_variant(evaluation_event)
+            evaluation_event.variant = self._variant_name_to_variant(
+                feature_flag, feature_flag.allocation.default_when_disabled
+            )
+            return
+
+        self._assign_variant(targeting_context, evaluation_event)
+
+    def _check_feature_base(self, feature_flag_id: str) -> Tuple[EvaluationEvent, bool]:
+        """
+        Determine if the feature flag is enabled for the given context.
+
+        :param str feature_flag_id: Name of the feature flag.
+        :return: The evaluation event and if the feature filters need to be checked.
+        :rtype: evaluation_event, bool
+        """
+        if self._copy is not self._configuration.get(FEATURE_MANAGEMENT_KEY):
+            self._cache = {}
+            self._copy = self._configuration.get(FEATURE_MANAGEMENT_KEY)
+
+        if not self._cache.get(feature_flag_id):
+            feature_flag = _get_feature_flag(self._configuration, feature_flag_id)
+            self._cache[feature_flag_id] = feature_flag
+        else:
+            feature_flag = self._cache.get(feature_flag_id)
+
+        evaluation_event = EvaluationEvent(feature_flag)
+        if not feature_flag:
+            logging.warning("Feature flag %s not found", feature_flag_id)
+            # Unknown feature flags are disabled by default
+            return evaluation_event, True
+
+        if not feature_flag.enabled:
+            # Feature flags that are disabled are always disabled
+            self._check_default_disabled_variant(evaluation_event)
+            if feature_flag.allocation:
+                variant_name = feature_flag.allocation.default_when_disabled
+                evaluation_event.variant = self._variant_name_to_variant(feature_flag, variant_name)
+            evaluation_event.feature = feature_flag
+
+            # If a feature flag is disabled and override can't enable it
+            evaluation_event.enabled = False
```

Comment:
I want to confirm that I understand this line.

This line exists for the case when the `default_when_disabled` variant has a `status_override=false`, since a feature flag will always be disabled when the "root" enabled flag is set to `false`. We don't want the disabled state of the feature flag to be overwritten.

Is my understanding correct?

Reply:
It would be when `status_override` of the default disabled variant is `true`. `is_enabled` should never return true when a feature flag has `enabled=False` but we still need to assign variants, reasons, and telemetry, and such.

Reply:
Yes it's when `status_override=true`, I made a mistake. But seems here to me this line exists because there is a side effect to `self._assign_default_disabled_variant()` as it calls `self.__assign_variant_override()`.

If we did `evaluation_event.reason = VariantAssignmentReason.DEFAULT_WHEN_DISABLED` at line 287, this line would not be necessary.